File: UVa 12516 cinema cola/main.py
'''
thought as dfs, actually a simple simulation problem
'''

import sys

sys.stdin = open('input.txt')


arr = [[False] * 100 for _ in range(27)]
seats = [[False] * 100 for _ in range(27)]
while True:
    rows, cols = map(int, raw_input().split())
    if (rows == 0 and cols == 0):
        break

    for _ in range(int(input())):
        rc, sign = raw_input().split()
        rowIn, colIn = ord(rc[0]) - 65, int(rc[1])
        if sign == '+':
            arr[rowIn][colIn] = True
        else:
            arr[rowIn][colIn - 1] = True

    for _ in range(int(input())):
        rc = raw_input()
        rowIn, colIn = ord(rc[0]) - 65, int(rc[1])
        seats[rowIn][colIn] = True

    canWe = True
    for r in range(rows):
        for c in range(1, cols + 1):
            if seats[r][c]:
                if not arr[r][c - 1]:
                    arr[r][c - 1] = True
                elif not arr[r][c]:
                    arr[r][c] = True
                else:
                    canWe = False
                    break
        if not canWe:
            break
    if (canWe):
        print 'YES'
    else:
        print 'NO'

File: UVa 12516 cinema cola/Main.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuffer sb = new StringBuffer("");
        boolean arr[][] = new boolean[27][100];
        boolean seats[][] = new boolean[27][100];
        while (true) {
            StringTokenizer str = new StringTokenizer(br.readLine());
            int rows = Integer.parseInt(str.nextToken());
            int cols = Integer.parseInt(str.nextToken());
            if (rows == 0 && cols == 0) {
                break;
            }
            for (int i = 0; i < 27; i++) {
                for (int j = 0; j < 100; j++) {
                    arr[i][j] = seats[i][j] = false;
                }
            }
            int alreadyIn = Integer.parseInt(br.readLine());
            for (int j = 0; j < alreadyIn; j++) {
                str = new StringTokenizer(br.readLine());
                StringBuilder temp = new StringBuilder(str.nextToken());
                int rowIn = temp.charAt(0) - 65;
                int colIn = Integer.parseInt(temp.substring(1));
                temp = new StringBuilder(str.nextToken());
                if (temp.charAt(0) == '-') {
                    arr[rowIn][colIn - 1] = true;
                } else {
                    arr[rowIn][colIn] = true;
                }
            }
            int myGroup = Integer.parseInt(br.readLine());
            for (int j = 0; j < myGroup; j++) {
                str = new StringTokenizer(br.readLine());
                StringBuilder temp = new StringBuilder(str.nextToken());
                int rowIn = temp.charAt(0) - 65;
                int colIn = Integer.parseInt(temp.substring(1));
                seats[rowIn][colIn] = true;
            }
            boolean canWe = true;
            for (int j = 0; j < rows; j++) {
                for (int z = 1; z < cols + 1; z++) {
                    if (seats[j][z]) {
                        if (!arr[j][z - 1]) {
                            arr[j][z - 1] = true;
                        } else if (!arr[j][z]) {
                            arr[j][z] = true;
                        } else {
                            canWe = false;
                            break;
                        }
                    }
                    if (!canWe)
                        break;
                }
            }
            if (canWe) {
                sb.append("YES").append("\n");
            } else {
                sb.append("NO").append("\n");
            }
        }
        System.out.print(sb);
    }
}